### PR TITLE
[SPARK-42657][CONNECT] Support to find and transfer client-side REPL classfiles to server as artifacts  

### PR DESCRIPTION
**Pull request by vicennial**
### What changes were proposed in this pull request?

This PR introduces the concept of a `ClassFinder` that is able to scrape the REPL output (either file-based or in-memory based) for generated class files.  The `ClassFinder` is registered during initialization of the REPL and aids in uploading the generated class files as artifacts to the Spark Connect server.

### Why are the changes needed?

To run UDFs which are defined on the client side REPL, we require a mechanism that can find the local REPL classfiles and then utilise the mechanism from https://issues.apache.org/jira/browse/SPARK-42653 to transfer them to the server as artifacts.  

### Does this PR introduce _any_ user-facing change?

Yes, users can now run UDFs on the default (ammonite) REPL with spark connect.

Input (in REPL):
```
class A(x: Int) { def get = x * 5 + 19 }
def dummyUdf(x: Int): Int = new A(x).get
val myUdf = udf(dummyUdf _)
spark.range(5).select(myUdf(col("id"))).as[Int].collect()
```

Output:
```
Array[Int] = Array(19, 24, 29, 34, 39)
```

### How was this patch tested?

Unit tests + E2E tests.

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -488,6 +488,14 @@ class SparkSession private[sql] (
   @scala.annotation.varargs
   def addArtifacts(uri: URI*): Unit = client.addArtifacts(uri)
 
+  /**
+   * Register a [[ClassFinder]] for dynamically generated classes.
+   *
+   * @since 3.4.0
+   */
+  @Experimental
```
[the remark this answers is not in the document]
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:496]: There is part of me that thinks we should be doing this as part of building the session, instead of adding it while running.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+          inputStream = ammoniteIn,
+          outputStream = ammoniteOut,
+          errorStream = errorStream)
+      }
+    }
+
+    executorService.submit(task)
+
+    // Run simple query to test REPL
+    val output = runCommandsInShell("spark.sql(\"select 1\").collect()")
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:66]: use a multi-line string that is easier to read. `"""spark.sql("select 1").collect()"""`

Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:66]: BTW just do this in a test. It is weird to have this in the beforeAll.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+package org.apache.spark.sql.application
+
+import java.io.{PipedInputStream, PipedOutputStream}
+import java.util.concurrent.{Executors, Semaphore, TimeUnit}
+
+import org.apache.commons.io.output.ByteArrayOutputStream
+
+import org.apache.spark.sql.connect.client.util.RemoteSparkSession
+
+class ReplE2ESuite extends RemoteSparkSession {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:26]: This should be beefed up quite a bit. @rednaxelafx @cloud-fan do we have a comprehensive set of tests somewhere?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:1]: Why this change?

**Review by hvanhovell — COMMENTED** [review]
Looks good overall, couple of small comments.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/SimpleSparkConnectService.scala [hunk 4]
```diff
@@ -38,7 +38,7 @@ private[sql] object SimpleSparkConnectService {
   private val stopCommand = "q"
 
   def main(args: Array[String]): Unit = {
-    val conf = new SparkConf()
+    val conf = new SparkConf().set("connect.test", "true")
```
[the remark this answers is not in the document]
Comment (vicennial) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/SimpleSparkConnectService.scala:41]: I tried this first but it breaks testing because it attempts to launch multiple servers (IIRC I was getting hit by port occupied errors)

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala
Comment (vicennial) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:1]: The `ReplE2ESuite` extends this suite and the port value is needed to passthrough to Ammonite

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (vicennial) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:496]: > There is part of me that thinks we should be doing this as part of building the session, instead of adding it while running.

I tried that initially but the `repl.sess` that we pass over in `spark.registerClassFinder(new AmmoniteClassFinder(repl.sess))` is only available once the REPL is instantiated

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala (hunk 3 above)
Comment (vicennial) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:26]: To clarify, you mean UDF specific tests right?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala (hunk 2 above)
Comment (vicennial) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:66]: Ack.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (vicennial) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:496]: > Also mark @Private?
Wrt the above comment, we can't since we register the class finder as part of the predef code. 
I'm not too well versed with Ammonite internals, I wonder if we can get the REPL session without having to go through predef code.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala (hunk 3 above)
Comment (vicennial) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:26]: That's odd. What were the test failures? 
Since it an E2E test, both the connect and server jars would need to be packaged before the tests. Could you try running the test after doing a `package`?

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -247,7 +247,10 @@ jobs:
     # Run the tests.
     - name: Run tests
       env: ${{ fromJSON(inputs.envs) }}
+      shell: 'script -q -e -c "bash {0}"'
       run: |
+        # Fix for TTY related issues when laaunching Ammonite REPL in tests.
```
Comment (grundprinzip) [inline .github/workflows/build_and_test.yml:252]: Typo

Comment (vicennial) [reply .github/workflows/build_and_test.yml]: Good eye, fixed!

**Review by hvanhovell — APPROVED** [review]
LGTM

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 6]
```diff
@@ -488,6 +488,14 @@ class SparkSession private[sql] (
   @scala.annotation.varargs
   def addArtifacts(uri: URI*): Unit = client.addArtifacts(uri)
 
+  /**
+   * Register a [[ClassFinder]] for dynamically generated classes.
+   *
+   * @since 3.4.0
```
Comment (pan3793) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:494]: should be 3.5.0

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala [hunk 7]
```diff
@@ -58,10 +58,13 @@ private[sql] class SparkConnectClient(
    * @return
    *   A [[proto.AnalyzePlanResponse]] from the Spark Connect server.
    */
-  def analyze(request: proto.AnalyzePlanRequest): proto.AnalyzePlanResponse =
+  def analyze(request: proto.AnalyzePlanRequest): proto.AnalyzePlanResponse = {
+    artifactManager.uploadAllClassFileArtifacts()
```
Comment (amaliujia) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:62]: Does this method cache previous scanned artifacts? If not maybe the overhead of calling analyze and execute becomes a bit large?

Comment (vicennial) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:62]: > Does this method cache previous scanned artifacts?

Good eye but not at the moment, no. However, the plan is to implement this later as an optimisation.

**Comment by LuciferYang** [comment]
Find an error related to ReplE2ESuite on [GA test task](https://pipelines.actions.githubusercontent.com/serviceHosts/c184045e-b556-4e78-b8ef-fb37b2eda9a3/_apis/pipelines/1/runs/69161/signedlogcontent/23?urlExpires=2023-04-18T10%3A18%3A24.5673026Z&urlSigningMethod=HMACV1&urlSignature=IZ4kWbB8mtkvxvyxojX3%2FxIz43j%2FVRKl7Ghp2Y52nnE%3D):

```
023-04-18T03:29:20.8938544Z [0m[[0m[0minfo[0m] [0m[0m[32mReplE2ESuite:[0m[0m
2023-04-18T03:29:24.5685770Z sh: 1: cannot open /dev/tty: No such device or address
...
2023-04-18T03:29:25.5551653Z sh: 1: cannot open /dev/tty: No such device or address
2023-04-18T03:29:25.5631256Z sh: 1: cannot create /dev/tty: No such device or address
...
2023-04-18T03:29:43.5473148Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.lang.RuntimeException: REPL Timed out while running command: [0m[0m
2023-04-18T03:29:43.5473697Z [0m[[0m[0minfo[0m] [0m[0m[31mclass A(x: Int) { def get = x * 5 + 19 }[0m[0m
2023-04-18T03:29:43.5474147Z [0m[[0m[0minfo[0m] [0m[0m[31mdef dummyUdf(x: Int): Int = new A(x).get[0m[0m
2023-04-18T03:29:43.5474578Z [0m[[0m[0minfo[0m] [0m[0m[31mval myUdf = udf(dummyUdf _)[0m[0m
2023-04-18T03:29:43.5480701Z [0m[[0m[0minfo[0m] [0m[0m[31mspark.range(5).select(myUdf(col("id"))).as[Int].collect()[0m[0m
2023-04-18T03:29:43.5481161Z [0m[[0m[0minfo[0m] [0m[0m[31m      [0m[0m
2023-04-18T03:29:43.5481539Z [0m[[0m[0minfo[0m] [0m[0m[31mConsole output: [0m[0m
2023-04-18T03:29:43.5482081Z [0m[[0m[0minfo[0m] [0m[0m[31mSpark session available as 'spark'.[0m[0m
2023-04-18T03:29:43.5484862Z [0m[[0m[0minfo[0m] [0m[0m[31m   _____                  __      ______                            __[0m[0m
2023-04-18T03:29:43.5488022Z [0m[[0m[0minfo[0m] [0m[0m[31m  / ___/____  ____ ______/ /__   / ____/___  ____  ____  ___  _____/ /_[0m[0m
2023-04-18T03:29:43.5491144Z [0m[[0m[0minfo[0m] [0m[0m[31m  \__ \/ __ \/ __ `/ ___/ //_/  / /   / __ \/ __ \/ __ \/ _ \/ ___/ __/[0m[0m
2023-04-18T03:29:43.5494244Z [0m[[0m[0minfo[0m] [0m[0m[31m ___/ / /_/ / /_/ / /  / ,<    / /___/ /_/ / / / / / / /  __/ /__/ /_[0m[0m
2023-04-18T03:29:43.5497421Z [0m[[0m[0minfo[0m] [0m[0m[31m/____/ .___/\__,_/_/  /_/|_|   \____/\____/_/ /_/_/ /_/\___/\___/\__/[0m[0m
2023-04-18T03:29:43.5500566Z [0m[[0m[0minfo[0m] [0m[0m[31m    /_/[0m[0m
2023-04-18T03:29:43.5502973Z sh: 1: cannot open /dev/tty: No such device or address
...
2023-04-18T03:29:43.7154910Z [0m[[0m[0minfo[0m] [0m[0m[31mError output: Compiling (synthetic)/ammonite/predef/ArgsPredef.sc[0m[0m
2023-04-18T03:29:43.7155688Z [0m[[0m[0minfo[0m] [0m[0m[31mCompiling /home/runner/work/spark/spark/connector/connect/client/jvm/(console)[0m[0m
2023-04-18T03:29:43.7156273Z [0m[[0m[0minfo[0m] [0m[0m[31mjava.lang.RuntimeException: Nonzero exit value: 2[0m[0m
2023-04-18T03:29:43.7156688Z [0m[[0m[0minfo[0m] [0m[0m[31m  scala.sys.package$.error(package.scala:30)[0m[0m
2023-04-18T03:29:43.7157121Z [0m[[0m[0minfo[0m] [0m[0m[31m  scala.sys.process.ProcessBuilderImpl$AbstractBuilder.slurp(ProcessBuilderImpl.scala:138)[0m[0m
2023-04-18T03:29:43.7157538Z [0m[[0m[0minfo[0m] [0m[0m[31m  scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang$bang(ProcessBuilderImpl.scala:108)[0m[0m
2023-04-18T03:29:43.7158104Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.terminal.TTY$.stty(Utils.scala:103)[0m[0m
2023-04-18T03:29:43.7158794Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.terminal.TTY$.withSttyOverride(Utils.scala:114)[0m[0m
2023-04-18T03:29:43.7160239Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.terminal.Terminal$.readLine(Terminal.scala:41)[0m[0m
2023-04-18T03:29:43.7162012Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.AmmoniteFrontEnd.readLine(AmmoniteFrontEnd.scala:133)[0m[0m
2023-04-18T03:29:43.7163249Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.AmmoniteFrontEnd.action(AmmoniteFrontEnd.scala:28)[0m[0m
2023-04-18T03:29:43.7163556Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Repl.$anonfun$action$4(Repl.scala:194)[0m[0m
2023-04-18T03:29:43.7163863Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Scoped.$anonfun$flatMap$1(Signaller.scala:45)[0m[0m
2023-04-18T03:29:43.7164158Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Signaller.apply(Signaller.scala:28)[0m[0m
2023-04-18T03:29:43.7171303Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Scoped.flatMap(Signaller.scala:45)[0m[0m
2023-04-18T03:29:43.7172151Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Scoped.flatMap$(Signaller.scala:45)[0m[0m
2023-04-18T03:29:43.7172894Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Signaller.flatMap(Signaller.scala:16)[0m[0m
2023-04-18T03:29:43.7174248Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Repl.$anonfun$action$2(Repl.scala:178)[0m[0m
2023-04-18T03:29:43.7176993Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.util.Catching.flatMap(Res.scala:115)[0m[0m
2023-04-18T03:29:43.7177811Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Repl.action(Repl.scala:170)[0m[0m
2023-04-18T03:29:43.7178715Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Repl.loop$1(Repl.scala:212)[0m[0m
2023-04-18T03:29:43.7179009Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.repl.Repl.run(Repl.scala:227)[0m[0m
2023-04-18T03:29:43.7179272Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.Main.$anonfun$run$1(Main.scala:236)[0m[0m
2023-04-18T03:29:43.7179543Z [0m[[0m[0minfo[0m] [0m[0m[31m  scala.Option.getOrElse(Option.scala:189)[0m[0m
2023-04-18T03:29:43.7179795Z [0m[[0m[0minfo[0m] [0m[0m[31m  ammonite.Main.run(Main.scala:224)[0m[0m
2023-04-18T03:29:43.7184102Z [0m[[0m[0minfo[0m] [0m[0m[31m  org.apache.spark.sql.application.ConnectRepl$.doMain(ConnectRepl.scala:100)[0m[0m
2023-04-18T03:29:43.7185205Z [0m[[0m[0minfo[0m] [0m[0m[31m  org.apache.spark.sql.application.ReplE2ESuite$$anon$1.run(ReplE2ESuite.scala:59)[0m[0m
2023-04-18T03:29:43.7185569Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[0m[0m
2023-04-18T03:29:43.7185885Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.util.concurrent.FutureTask.run(FutureTask.java:266)[0m[0m
2023-04-18T03:29:43.7191750Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[0m[0m
2023-04-18T03:29:43.7192249Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[0m[0m
2023-04-18T03:29:43.7192662Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.lang.Thread.run(Thread.java:750)[0m[0m
```

I am currently unsure why this error occurred

**Comment by vicennial** [comment]
@LuciferYang It looks related to the TTY issues (https://github.com/com-lihaoyi/Ammonite/issues/276) we were hitting in the CI pipelines earlier. (`2023-04-18T03:29:24.5685770Z sh: 1: cannot open /dev/tty: No such device or address` in the logs is the indicator) 

[These](https://github.com/apache/spark/pull/40675/files#diff-48c0ee97c53013d18d6bbae44648f7fab9af2e0bf5b0dc1ca761e18ec5c478f2R250-R253) changes (taken from [here](https://github.com/com-lihaoyi/Ammonite/issues/276#issuecomment-439273906)) (helped us mitigate the issue for the CI pipelines but perhaps the GA test pipeline uses a different config? Any idea about the pipeline config @HyukjinKwon?

**Comment by HyukjinKwon** [comment]
Is this just being flaky? All GA jobs share the same thing (that you fixed)

**Comment by zhenlineo** [comment]
@HyukjinKwon I would not think it is flaky. It is probably the GA builds also missing the TTY setup in the build steps. So we need to copy the following changes to the steps:
```
      run: |
        # Fix for TTY related issues when launching the Ammonite REPL in tests.
        export TERM=vt100 && script -qfc 'echo exit | amm -s' && rm typescript
```
https://github.com/apache/spark/pull/40675/files#diff-48c0ee97c53013d18d6bbae44648f7fab9af2e0bf5b0dc1ca761e18ec5c478f2R250-R253

**Comment by HyukjinKwon** [comment]
@LuciferYang which GitHub job failed? The link seems stale now so I can't see. It is a scheduled build?

**Comment by LuciferYang** [comment]
> @LuciferYang which GitHub job failed? The link seems stale now so I can't see. It is a scheduled build?

I found this issue from GA task of user pr

The last failed build on https://github.com/apache/spark/pull/40783  also had this issue, I don't know if @rangadi  can provide a new link

**Comment by LuciferYang** [comment]
@HyukjinKwon https://github.com/rangadi/spark/actions/runs/4737137341/jobs/8409588363 this one

**Comment by HyukjinKwon** [comment]
Ah, i think it just happened because of unsynced fork. It should be fixed if their fork is synced to the lastest master branch.

**Comment by LuciferYang** [comment]
Hmm... GA should merge the current pr to the latest master before testing, right? 

It's okay, Let's wait and see if there are any new cases happening.

**Comment by pan3793** [comment]
> @HyukjinKwon I would not think it is flaky. It is probably the GA builds also missing the TTY setup in the build steps. So we need to copy the following changes to the steps:
> 
> ```
>       run: |
>         # Fix for TTY related issues when launching the Ammonite REPL in tests.
>         export TERM=vt100 && script -qfc 'echo exit | amm -s' && rm typescript
> ```
> 
> https://github.com/apache/spark/pull/40675/files#diff-48c0ee97c53013d18d6bbae44648f7fab9af2e0bf5b0dc1ca761e18ec5c478f2R250-R253

@zhenlineo is there some reference about such magic command? better to briefly explain the mechanism or leave a link in comments

**Comment by HyukjinKwon** [comment]
> Hmm... GA should merge the current pr to the latest master before testing, right?

Yes except the workflow file. For the changes in workflow, they have to manually update to the latest master branch.

**Comment by LuciferYang** [comment]
> > Hmm... GA should merge the current pr to the latest master before testing, right?
> 
> Yes except the workflow file. For the changes in workflow, they have to manually update to the latest master branch.

Got it ~

**Comment by LuciferYang** [comment]
@vicennial I found `ReplE2ESuite` always failed in Java 17 GA daily test:

- https://github.com/apache/spark/actions/runs/4726264540/jobs/8385681548
- https://github.com/apache/spark/actions/runs/4737365554/jobs/8410097712
- https://github.com/apache/spark/actions/runs/4748319019/jobs/8434392414
- https://github.com/apache/spark/actions/runs/4759278349/jobs/8458399201

image

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    }
+    getCleanString(ammoniteOut)
+  }
+
+  def assertContains(message: String, output: String): Unit = {
+    val isContain = output.contains(message)
+    assert(isContain, "Ammonite output did not contain '" + message + "':\n" + output)
+  }
+
+  test("Simple query") {
```
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:97]: This test consistently fails with JDK 17:

```
[info] ReplE2ESuite:
[info] - Simple query *** FAILED *** (10 seconds, 4 milliseconds)
[info]   java.lang.RuntimeException: REPL Timed out while running command: 
[info] spark.sql("select 1").collect()
[info]       
[info] Console output: 
[info] Error output: Compiling (synthetic)/ammonite/predef/ArgsPredef.sc
[info]   at org.apache.spark.sql.application.ReplE2ESuite.runCommandsInShell(ReplE2ESuite.scala:87)
[info]   at org.apache.spark.sql.application.ReplE2ESuite.$anonfun$new$1(ReplE2ESuite.scala:102)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[info]   at org.scalatest.funsuite.AnyFunSuite.withFixture(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
```
https://github.com/apache/spark/actions/runs/4780630672/jobs/8498505928#step:9:4647
https://github.com/apache/spark/actions/runs/4774942961/jobs/8488946907
https://github.com/apache/spark/actions/runs/4769162286/jobs/8479293802
https://github.com/apache/spark/actions/runs/4759278349/jobs/8458399201
https://github.com/apache/spark/actions/runs/4748319019/jobs/8434392414

Comment (vicennial) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:97]: @HyukjinKwon Verified that the fix works locally, ptal at PR: https://github.com/apache/spark/pull/40948

Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:97]: Thx!

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala:97]: https://github.com/apache/spark/actions/runs/4813779407

Yesterday's test passed, thanks ~
